numba/typed/typedobjectutils.py

```diff
@@ -197,3 +197,78 @@ def build_wrapper(fn):
     builder.ret(context.get_constant(types.int32, -1))
 
     return equal_fn
+
+
+def _get_primitive_equal(context, module, datamodel, element_type):
+    assert not datamodel.contains_nrt_meminfo()
+
+    data_ptr_ty = ir.IntType(8).as_pointer()
+    equal_fnty = ir.FunctionType(ir.IntType(32), [data_ptr_ty, data_ptr_ty])
+    equal_fn = cgutils.get_or_insert_function(
+        module,
+        equal_fnty,
+        name='.numba_{}.{}_equal'.format(
+            context.fndesc.mangled_name, element_type),
+    )
+    builder = ir.IRBuilder(equal_fn.append_basic_block())
+    lhs_ptr = builder.bitcast(equal_fn.args[0], datamodel.be_type.as_pointer())
+    lhs = builder.load(lhs_ptr)
+    rhs_ptr = builder.bitcast(equal_fn.args[1], datamodel.be_type.as_pointer())
+    rhs = builder.load(rhs_ptr)
+    if datamodel.fe_type.signed:
+        res = builder.icmp_signed('==', lhs, rhs)
+    else:
+        res = builder.icmp_unsigned('==', lhs, rhs)
+    with builder.if_else(res) as (then, orelse):
+        with then:
+            builder.ret(context.get_constant(types.int32, 1))
+        with orelse:
+            builder.ret(context.get_constant(types.int32, 0))
+
+    builder.ret(context.get_constant(types.int32, -1))
```

Comment:
You can return `res` instead, as it will be an integer holding either `0` or `1`.

Reply:
In fact, I tried to directly `builder.ret(res)`, but I will get a compilation error, to complain I messed up sth with `i32`. 
Need to dig more.

Reply:
Oh, I realized if I change the return value type of `key_equal` in C declaration to `bool`, perhaps I don't need to add this branch. Here is a snippet of generated IR:
```llvm
; Function Attrs: mustprogress nofree norecurse nosync nounwind readonly willreturn
define i32 @.numba__ZN5numba5typed10dictobject13impl_new_dict12_3clocals_3e3impB3v15B42c8tJTIcFKzyF2ILShI4CrgQElcaMVJhEEUSJJgA_3dE16class_28int64_2916class_28int64_29x.dict_key_equal(i8* nocapture readonly %.1, i8* nocapture readonly %.2) #0 {
.4:
  %.5 = bitcast i8* %.1 to i64*
  %.6 = load i64, i64* %.5, align 8
  %.7 = bitcast i8* %.2 to i64*
  %.8 = load i64, i64* %.7, align 8
  %.9 = icmp eq i64 %.6, %.8
  %. = zext i1 %.9 to i32
  ret i32 %.
}
```

Reply:
But in the usage side, it wants to get 3 types of results as the comment pointed out, so I guess I can't directly return a `bool`. Here is the comment for `key_equal`:
```c
/* Returns -1 for error; 0 for not equal; 1 for equal */
```